### DISCUSSION TITLE
Bug-Memory will not be released after the query!!!

### DISCUSSION THREAD

**Discussion by kqkdChen** [trinodb/trino#13792] [General]
Trino Version: 366
Run On Kubernetes：v1.19.4 
Use helm to arrange containers
```
Coordinator configuration：
query.max-memory-per-node: 11GB
query.max-total-memory-per-node: 21GB
query.max-memory: 44GB
query.max-total-memory: 88GB
JVM-heapSize: 52GB
# hardware information
work-nums: 4
Server memory：64G
kubernetes Pod-memory-limit: 54GB
```
![coorconfig](https://user-images.githubusercontent.com/41674301/147568790-1b7225b7-6d0f-4448-9165-bc0ae8810943.jpg)

work configuration:
![workconfig](https://user-images.githubusercontent.com/41674301/147568826-66c33078-28ea-491d-bfd4-564969581359.jpg)

JVM configuration:
![jvmconfig](https://user-images.githubusercontent.com/41674301/147568921-1042b81a-b114-4864-8ae0-174692fac3bc.jpg)

As shown in the figure, the memory after the query will not be released until the memory is full and then the pod will be killed

![memory](https://user-images.githubusercontent.com/41674301/147569920-b30c2f5c-ecc0-4ccc-81bc-a1fc94a3f690.jpg)

**Comment by sondn98** [comment]
I'm also facing with this problem, any idea here?

**Comment by kqkdChen** [comment]
> I'm also facing with this problem, any idea here?
> 
> 我也面临着这个问题，有什么想法吗？

Using `ShenandoahGC` can solve this problem, 2T data query 1000 times the trino cluster performance is very stable, the only defect is that the performance drops by about 10%

**Comment by hashhar** [comment]
This is how the JVM is supposed to work. Heap memory is not released back to the OS automatically. You should inspect whether there is remaining memory in heap instead of looking at memory from the OS.

Closing, feel free to continue discussion.
